src/language-mode.coffee

```diff
@@ -246,23 +246,30 @@ class LanguageMode
     iterator.next()
     scopeDescriptor = new ScopeDescriptor(scopes: iterator.getScopes())
 
+    increaseNextIndentRegex = @increaseNextIndentRegexForScopeDescriptor(scopeDescriptor)
+    decreaseNextIndentRegex = @decreaseNextIndentRegexForScopeDescriptor(scopeDescriptor)
+    decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
+
     currentIndentLevel = @editor.indentationForBufferRow(bufferRow)
-    return currentIndentLevel unless increaseIndentRegex = @increaseIndentRegexForScopeDescriptor(scopeDescriptor)
+    return currentIndentLevel unless increaseNextIndentRegex
 
-    if options?.skipBlankLines ? true
+    bufferLine = @buffer.lineForRow(bufferRow)
+    if options?.skipBlankLines ? true or (decreaseIndentRegex?.testSync(bufferLine) and not @editor.isBufferRowCommented(bufferRow))
```

Comment:
:zap: This is looking very clean now.

I think we still need to respect the `skipBlankLines` option here: if it is set to false, we should only consider the previous row, and not search upward for the previous *non-blank* row. The reason is that when inserting a newline by pressing `enter`, we want to avoid altering the indentation based on 'far-away' lines of the buffer.

For example, suppose that somewhere above the cursor, there's a line whose indentation doesn't match the tab width. This happens:

![indent](https://cloud.githubusercontent.com/assets/326587/8288688/984c50fe-18cc-11e5-9e4e-dfc34e070d52.gif)

Even though the 'correct' indentation is being suggested, based on the settings, it can feel unintuitive to the user because they're just trying to insert a newline.

Thinking about alternative ways to make your tests pass...